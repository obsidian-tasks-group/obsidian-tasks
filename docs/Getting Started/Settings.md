---
publish: true
---

# Settings

## Restart after changing settings

Regrettably, many of the settings for customising Tasks require one of the following to be done, for the new options to take effect:

- Obsidian to be restarted,
- or the vault to be closed and re-opened.

The Tasks Settings pane warns of this:

![Changing any settings requires a restart of obsidian](../images/settings-restart-after-change.png)

## Available settings

As the number of options in Tasks has grown, we have documented each setting in the page for the relevant topic.

For convenience, here is a list of all those documentation pages (in the order the options appear in the Tasks Settings pane):

- [[About Task Formats#Selecting the task format]]
- [[Global Filter]]
- [[Global Query]]
- [[Status Settings]]
- [[Dates#Date-tracking settings]]
- [[Recurring Tasks#Order of the new task]]
- [[Auto-Suggest]]
- [[Create or edit Task#Turning off keyboard shortcuts]]
