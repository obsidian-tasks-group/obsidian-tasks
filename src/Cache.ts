import { MetadataCache, TAbstractFile, TFile, Vault } from 'obsidian';
import type { CachedMetadata, EventRef } from 'obsidian';
import type { HeadingCache, ListItemCache, SectionCache } from 'obsidian';
import { Mutex } from 'async-mutex';

import { Task } from './Task';
import type { Events } from './Events';

export enum State {
    Cold = 'Cold',
    Initializing = 'Initializing',
    Warm = 'Warm',
}

export class Cache {
    private readonly metadataCache: MetadataCache;
    private readonly metadataCacheEventReferences: EventRef[];
    private readonly vault: Vault;
    private readonly vaultEventReferences: EventRef[];
    private readonly events: Events;
    private readonly eventsEventReferences: EventRef[];

    private readonly tasksMutex: Mutex;
    private state: State;
    private tasks: Task[];

    /**
     * We cannot know if this class will be instantiated because obsidian started
     * or because the plugin was activated later. This means we have to load the
     * whole vault once after the first metadata cache resolve to ensure that we
     * load the entire vault in case obsidian is starting up. In the case of
     * obsidian starting, the task cache's initial load would end up with 0 tasks,
     * as the metadata cache would still be empty.
     */
    private loadedAfterFirstResolve: boolean;

    constructor({
        metadataCache,
        vault,
        events,
    }: {
        metadataCache: MetadataCache;
        vault: Vault;
        events: Events;
    }) {
        this.metadataCache = metadataCache;
        this.metadataCacheEventReferences = [];
        this.vault = vault;
        this.vaultEventReferences = [];
        this.events = events;
        this.eventsEventReferences = [];

        this.tasksMutex = new Mutex();
        this.state = State.Cold;
        this.tasks = [];

        this.loadedAfterFirstResolve = false;

        this.subscribeToCache();
        this.subscribeToVault();
        this.subscribeToEvents();

        this.loadVault();
    }

    public unload(): void {
        for (const eventReference of this.metadataCacheEventReferences) {
            this.metadataCache.offref(eventReference);
        }

        for (const eventReference of this.vaultEventReferences) {
            this.vault.offref(eventReference);
        }

        for (const eventReference of this.eventsEventReferences) {
            this.events.off(eventReference);
        }
    }

    public getTasks(): Task[] {
        return this.tasks;
    }

    public getState(): State {
        return this.state;
    }

    private notifySubscribers(): void {
        this.events.triggerCacheUpdate({
            tasks: this.tasks,
            state: this.state,
        });
    }

    private subscribeToCache(): void {
        const resolvedEventeReference = this.metadataCache.on(
            'resolved',
            async () => {
                // Resolved fires on every change.
                // We only want to initialize if we haven't already.
                if (!this.loadedAfterFirstResolve) {
                    this.loadedAfterFirstResolve = true;
                    this.loadVault();
                }
            },
        );
        this.metadataCacheEventReferences.push(resolvedEventeReference);

        // Does not fire when starting up obsidian and only works for changes.
        const changedEventReference = this.metadataCache.on(
            'changed',
            (file: TFile) => {
                this.tasksMutex.runExclusive(() => {
                    this.indexFile(file);
                });
            },
        );
        this.metadataCacheEventReferences.push(changedEventReference);
    }

    private subscribeToVault(): void {
        const createdEventReference = this.vault.on(
            'create',
            (file: TAbstractFile) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.indexFile(file);
                });
            },
        );
        this.vaultEventReferences.push(createdEventReference);

        const deletedEventReference = this.vault.on(
            'delete',
            (file: TAbstractFile) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.tasks = this.tasks.filter((task: Task) => {
                        return task.path !== file.path;
                    });

                    this.notifySubscribers();
                });
            },
        );
        this.vaultEventReferences.push(deletedEventReference);

        const renamedEventReference = this.vault.on(
            'rename',
            (file: TAbstractFile, oldPath: string) => {
                if (!(file instanceof TFile)) {
                    return;
                }

                this.tasksMutex.runExclusive(() => {
                    this.tasks = this.tasks.map((task: Task): Task => {
                        if (task.path === oldPath) {
                            return new Task({ ...task, path: file.path });
                        } else {
                            return task;
                        }
                    });

                    this.notifySubscribers();
                });
            },
        );
        this.vaultEventReferences.push(renamedEventReference);
    }

    private subscribeToEvents(): void {
        const requestReference = this.events.onRequestCacheUpdate((handler) => {
            handler({ tasks: this.tasks, state: this.state });
        });
        this.eventsEventReferences.push(requestReference);
    }

    private loadVault(): Promise<void> {
        return this.tasksMutex.runExclusive(async () => {
            this.state = State.Initializing;
            await Promise.all(
                this.vault.getMarkdownFiles().map((file: TFile) => {
                    return this.indexFile(file);
                }),
            );
            this.state = State.Warm;
            // Notify that the cache is now warm:
            this.notifySubscribers();
        });
    }

    private async indexFile(file: TFile): Promise<void> {
        const fileCache = this.metadataCache.getFileCache(file);
        if (fileCache === null || fileCache === undefined) {
            return;
        }

        const oldTasks = this.tasks.filter((task: Task) => {
            return task.path === file.path;
        });

        const listItems = fileCache.listItems;
        // When there is no list items cache, there are no tasks.
        // Still continue to notify watchers of removal.

        let newTasks: Task[] = [];
        if (listItems !== undefined) {
            // Only read the file and process for tasks if there are list items.
            const fileContent = await this.vault.cachedRead(file);
            newTasks = Cache.getTasksFromFileContent(
                fileContent,
                listItems,
                fileCache,
                file,
            );
        }

        // If there are no changes in any of the tasks, there's
        // nothing to do, so just return.
        if (Task.tasksListsIdentical(oldTasks, newTasks)) {
            // This code kept for now, to allow for debugging during development.
            // It is too verbose to release to users.
            // if (this.getState() == State.Warm) {
            //     console.debug(`Tasks unchanged in ${file.path}`);
            // }
            return;
        }

        if (this.getState() == State.Warm) {
            console.debug(
                `At least one task, its line number or its heading has changed in ${file.path}: triggering a refresh of all active Tasks blocks in Live Preview and Reading mode views.`,
            );
        }

        // Remove all tasks from this file from the cache before
        // adding the ones that are currently in the file.
        this.tasks = this.tasks.filter((task: Task) => {
            return task.path !== file.path;
        });

        this.tasks.push(...newTasks);

        // All updated, inform our subscribers.
        this.notifySubscribers();
    }

    private static getTasksFromFileContent(
        fileContent: string,
        listItems: ListItemCache[],
        fileCache: CachedMetadata,
        file: TFile,
    ): Task[] {
        const tasks: Task[] = [];
        const fileLines = fileContent.split('\n');

        // We want to store section information with every task so
        // that we can use that when we post process the markdown
        // rendered lists.
        let currentSection: SectionCache | null = null;
        let sectionIndex = 0;
        for (const listItem of listItems) {
            if (listItem.task !== undefined) {
                if (
                    currentSection === null ||
                    currentSection.position.end.line <
                        listItem.position.start.line
                ) {
                    // We went past the current section (or this is the first task).
                    // Find the section that is relevant for this task and the following of the same section.
                    currentSection = Cache.getSection(
                        listItem.position.start.line,
                        fileCache.sections,
                    );
                    sectionIndex = 0;
                }

                if (currentSection === null) {
                    // Cannot process a task without a section.
                    continue;
                }

                const line = fileLines[listItem.position.start.line];
                const task = Task.fromLine({
                    line,
                    path: file.path,
                    sectionStart: currentSection.position.start.line,
                    sectionIndex,
                    precedingHeader: Cache.getPrecedingHeader(
                        listItem.position.start.line,
                        fileCache.headings,
                    ),
                });

                if (task !== null) {
                    sectionIndex++;
                    tasks.push(task);
                }
            }
        }

        return tasks;
    }

    private static getSection(
        lineNumberTask: number,
        sections: SectionCache[] | undefined,
    ): SectionCache | null {
        if (sections === undefined) {
            return null;
        }

        for (const section of sections) {
            if (
                section.position.start.line <= lineNumberTask &&
                section.position.end.line >= lineNumberTask
            ) {
                return section;
            }
        }

        return null;
    }

    private static getPrecedingHeader(
        lineNumberTask: number,
        headings: HeadingCache[] | undefined,
    ): string | null {
        if (headings === undefined) {
            return null;
        }

        let precedingHeader: string | null = null;

        for (const heading of headings) {
            if (heading.position.start.line > lineNumberTask) {
                return precedingHeader;
            }
            precedingHeader = heading.heading;
        }
        return precedingHeader;
    }
}
