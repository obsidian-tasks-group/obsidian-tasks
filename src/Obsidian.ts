import {
    App,
    EventRef,
    MarkdownPostProcessor,
    MarkdownPostProcessorContext,
    MarkdownView,
    Plugin,
    TAbstractFile,
    TFile,
    Vault,
    Workspace,
} from 'obsidian';

export class Obsidian {
    private readonly plugin: Plugin;
    private readonly eventReferences: EventRef[] = [];

    constructor({ plugin }: { plugin: Plugin }) {
        this.plugin = plugin;
    }

    public unload(): void {
        for (const eventRef of this.eventReferences) {
            this.vault.offref(eventRef);
        }
    }

    private get app(): App {
        return this.plugin.app;
    }

    private get vault(): Vault {
        return this.app.vault;
    }

    private get workspace(): Workspace {
        return this.app.workspace;
    }

    public get activeFilePath(): string | undefined {
        return (this.workspace.activeLeaf.view as any)?.file?.path;
    }

    public get activeMarkdownView(): MarkdownView | undefined {
        if (this.workspace.activeLeaf === undefined) {
            return undefined;
        }

        const activeLeaf = this.workspace.activeLeaf;
        if (!(activeLeaf.view instanceof MarkdownView)) {
            return undefined;
        }

        return activeLeaf.view;
    }

    public get editor(): CodeMirror.Editor {
        return (this.workspace.activeLeaf.view as any).sourceMode
            .cmEditor as CodeMirror.Editor;
    }

    public getMarkdownFilePaths(): string[] {
        return this.vault.getMarkdownFiles().map((file) => file.path);
    }

    public async readLines({ path }: { path: string }): Promise<string[]> {
        const file = this.vault.getAbstractFileByPath(path);
        const fileContent = await this.vault.read(file as TFile);
        const fileLines = fileContent.split('\n');

        return fileLines;
    }

    public writeLines({
        path,
        lines,
    }: {
        path: string;
        lines: string[];
    }): Promise<void> {
        const file = this.vault.getAbstractFileByPath(path);
        const fileContent = lines.join('\n');
        return this.vault.modify(file as TFile, fileContent);
    }

    public addCommand(command: {
        id: string;
        name: string;
        checkCallback: (checking: boolean) => void | boolean;
    }) {
        this.plugin.addCommand(command);
    }

    public subscribeToLayoutReadyEvent(func: () => void): void {
        this.workspace.onLayoutReady(func);
    }

    public subscribeToCreation(func: (path: string) => Promise<void>) {
        const eventRef = this.vault.on('create', (file: TAbstractFile) => {
            func(file.path);
        });

        this.eventReferences.push(eventRef);
    }

    public subscribeToModification(func: (path: string) => Promise<void>) {
        const eventRef = this.vault.on('modify', (file: TAbstractFile) => {
            func(file.path);
        });

        this.eventReferences.push(eventRef);
    }

    public subscribeToDeletion(func: (path: string) => Promise<void>) {
        const eventRef = this.vault.on('delete', (file: TAbstractFile) => {
            func(file.path);
        });

        this.eventReferences.push(eventRef);
    }

    public subscribeToRenaming(
        func: (oldPath: string, newPath: string) => Promise<void>,
    ) {
        const eventRef = this.vault.on(
            'rename',
            (file: TAbstractFile, oldPath: string) => {
                func(oldPath, file.path);
            },
        );

        this.eventReferences.push(eventRef);
    }

    public registerMarkdownPostProcessor(handler: MarkdownPostProcessor): void {
        this.plugin.registerMarkdownPostProcessor(handler);
    }

    public registerCodeBlockPostProcessor(
        handler: (
            source: string,
            el: HTMLElement,
            ctx: MarkdownPostProcessorContext,
        ) => Promise<any> | void,
    ): void {
        this.plugin.registerMarkdownCodeBlockProcessor('tasks', handler);
    }
}
