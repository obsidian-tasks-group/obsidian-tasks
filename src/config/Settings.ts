import { Feature } from './Feature';
import type { FeatureFlag } from './Feature';

export interface Settings {
    globalFilter: string;
    removeGlobalFilter: boolean;
    setDoneDate: boolean;
    autoSuggestInEditor: boolean;
    autoSuggestMinMatch: number;
    autoSuggestMaxItems: number;

    // Collection of feature flag IDs and their state.
    features: FeatureFlag;
}

const defaultSettings: Settings = {
    globalFilter: '',
    removeGlobalFilter: false,
    setDoneDate: true,
    autoSuggestInEditor: true,
    autoSuggestMinMatch: 0,
    autoSuggestMaxItems: 6,
    features: Feature.settingsFlags,
};

let settings: Settings = { ...defaultSettings };

/**
 * Returns the current settings as a object, it will also check and
 * update the flags to make sure they are all shown in the data.json
 * file. Exposure via the settings UI is optional.
 *
 * @export
 * @returns true if the feature is enabled.
 */
export const getSettings = (): Settings => {
    // Check to see if there is a new flag and if so add it to the users settings.
    for (const flag in Feature.settingsFlags) {
        if (settings.features[flag] === undefined) {
            settings.features[flag] = Feature.settingsFlags[flag];
        }
    }

    return { ...settings };
};

export const updateSettings = (newSettings: Partial<Settings>): Settings => {
    settings = { ...settings, ...newSettings };

    return getSettings();
};

export const resetSettings = (): Settings => {
    return updateSettings(defaultSettings);
};

/**
 * Returns the enabled state of the feature from settings.
 *
 * @export
 * @param internalName the internal name of the feature.
 * @returns true if the feature is enabled.
 */
export const isFeatureEnabled = (internalName: string): boolean => {
    return settings.features[internalName] ?? false;
};

/**
 * enables toggling the feature and returning the current collection with state.
 *
 * @export
 * @param internalName the internal name of the feature.
 * @param enabled the expected state of the feature.
 * @returns the features with the specified feature toggled.
 */
export const toggleFeature = (
    internalName: string,
    enabled: boolean,
): FeatureFlag => {
    settings.features[internalName] = enabled;
    return settings.features;
};
