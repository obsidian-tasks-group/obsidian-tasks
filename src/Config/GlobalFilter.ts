import * as RegExpTools from '../lib/RegExpTools';
import { getSettings, updateSettings } from './Settings';

export class GlobalFilter {
    static empty = '';

    static get(): string {
        const { globalFilter } = getSettings();
        return globalFilter;
    }

    static set(value: string) {
        updateSettings({ globalFilter: value });
    }

    static reset() {
        updateSettings({ globalFilter: GlobalFilter.empty });
    }

    static isEmpty(): boolean {
        return GlobalFilter.get() === GlobalFilter.empty;
    }

    static equals(tag: string): boolean {
        return GlobalFilter.get() === tag;
    }

    static includedIn(description: string): boolean {
        const globalFilter = GlobalFilter.get();
        return description.includes(globalFilter);
    }

    static prependTo(description: string): string {
        return GlobalFilter.get() + ' ' + description;
    }

    static addGlobalFilterToDescriptionDependingOnSettings(description: string): string {
        if (GlobalFilter.shouldAddGlobalFilter(description)) {
            return GlobalFilter.prependTo(description);
        } else {
            return description;
        }
    }

    private static shouldAddGlobalFilter(description: string): boolean {
        const { autoInsertGlobalFilter } = getSettings();

        return !GlobalFilter.isEmpty() && autoInsertGlobalFilter && !GlobalFilter.includedIn(description);
    }

    static removeAsWordFromDependingOnSettings(description: string): string {
        const { removeGlobalFilter } = getSettings();
        if (removeGlobalFilter) {
            return GlobalFilter.removeAsWordFrom(description);
        }

        return description;
    }

    /**
     * Search for the global filter for the purpose of removing it from the description, but do so only
     * if it is a separate word (preceding the beginning of line or a space and followed by the end of line
     * or a space), because we don't want to cut-off nested tags like #task/subtag.
     * If the global filter exists as part of a nested tag, we keep it untouched.
     */
    static removeAsWordFrom(description: string): string {
        if (GlobalFilter.isEmpty()) {
            return description;
        }

        // This matches the global filter (after escaping it) only when it's a complete word
        const theRegExp = RegExp('(^|\\s)' + RegExpTools.escapeRegExp(GlobalFilter.get()) + '($|\\s)', 'ug');

        if (description.search(theRegExp) > -1) {
            description = description.replace(theRegExp, '$1$2').replace('  ', ' ').trim();
        }

        return description;
    }

    static removeAsSubstringFrom(description: string): string {
        const globalFilter = GlobalFilter.get();
        return description.replace(globalFilter, '').trim();
    }
}
