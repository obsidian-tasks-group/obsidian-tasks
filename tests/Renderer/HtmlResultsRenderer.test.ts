/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import type { Task } from 'Task/Task';
import { GlobalFilter } from '../../src/Config/GlobalFilter';
import { State } from '../../src/Obsidian/Cache';
import { HtmlQueryResultsRenderer } from '../../src/Renderer/HtmlQueryResultsRenderer';
import { TasksFile } from '../../src/Scripting/TasksFile';
import { verifyWithFileExtension } from '../TestingTools/ApprovalTestHelpers';
import { prettifyHTML } from '../TestingTools/HTMLHelpers';
import { TaskBuilder } from '../TestingTools/TaskBuilder';
import { toMarkdown } from '../TestingTools/TestHelpers';
import { resetSettings } from '../../src/Config/Settings';
import { mockApp } from '../__mocks__/obsidian';
import { readTasksFromSimulatedFile } from '../Obsidian/SimulatedFile';
import { Query } from '../../src/Query/Query';
import { mockHTMLRenderer } from './RenderingTestHelpers';

window.moment = moment;

beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2023-07-05'));
});

afterEach(() => {
    jest.useRealTimers();
    GlobalFilter.getInstance().reset();
    resetSettings();
});

function makeHtmlResultsRenderer(source: string, tasksFile: TasksFile) {
    const query = new Query(source, tasksFile);
    return new HtmlQueryResultsRenderer(
        source,
        tasksFile,
        query,
        () => Promise.resolve(),
        null,
        mockApp,
        mockHTMLRenderer,
    );
}

describe('HtmlResultsRenderer tests', () => {
    async function verifyRenderedTasksHTML(allTasks: Task[], source: string = '') {
        const renderer = makeHtmlResultsRenderer(source, new TasksFile('query.md'));
        const queryRendererParameters = {
            allTasks,
            allMarkdownFiles: [],
            backlinksClickHandler: () => Promise.resolve(),
            backlinksMousedownHandler: () => Promise.resolve(),
            editTaskPencilClickHandler: () => Promise.resolve(),
        };
        const container = document.createElement('div');

        await renderer.render(State.Warm, allTasks, container, queryRendererParameters);

        const taskAsMarkdown = `<!--
${toMarkdown(allTasks)}
-->\n\n`;

        const prettyHTML = prettifyHTML(container.outerHTML);
        verifyWithFileExtension(taskAsMarkdown + prettyHTML, 'html');
    }

    it('fully populated task', async () => {
        const allTasks = [TaskBuilder.createFullyPopulatedTask()];
        await verifyRenderedTasksHTML(allTasks, 'show urgency');
    });

    it('fully populated task - short mode', async () => {
        const allTasks = [TaskBuilder.createFullyPopulatedTask()];
        await verifyRenderedTasksHTML(allTasks, 'show urgency\nshort mode');
    });

    const showTree = 'show tree\n';
    const hideTree = 'hide tree\n';

    it('parent-child items hidden', async () => {
        const allTasks = readTasksFromSimulatedFile('inheritance_rendering_sample');
        await verifyRenderedTasksHTML(allTasks, hideTree + 'sort by function task.lineNumber');
    });

    it('parent-child items', async () => {
        const allTasks = readTasksFromSimulatedFile('inheritance_rendering_sample');
        await verifyRenderedTasksHTML(allTasks, showTree + 'sort by function task.lineNumber');
    });

    it('parent-child items reverse sorted', async () => {
        const allTasks = readTasksFromSimulatedFile('inheritance_rendering_sample');
        await verifyRenderedTasksHTML(allTasks, showTree + 'sort by function reverse task.lineNumber');
    });

    it('should render tasks without their parents', async () => {
        // example chosen to match subtasks whose parents do not match the query
        const allTasks = readTasksFromSimulatedFile('inheritance_task_2listitem_3task');
        await verifyRenderedTasksHTML(allTasks, showTree + 'description includes grandchild');
    });

    it('should render non task check box when global filter is enabled', async () => {
        GlobalFilter.getInstance().set('#task');
        const allTasks = readTasksFromSimulatedFile('inheritance_non_task_child');
        await verifyRenderedTasksHTML(allTasks, showTree);
    });
});

describe('HtmlResultsRenderer - internal heading links', () => {
    let tasksByHeading: Record<string, Task>;

    beforeAll(() => {
        const allTasks = readTasksFromSimulatedFile('internal_heading_links');

        tasksByHeading = allTasks.reduce((acc, task) => {
            const heading = task.taskLocation.precedingHeader ?? '';

            // For now, the test design only supports one task per heading, so make it an error
            // if there are multiple tasks in this heading:
            if (acc[heading]) {
                throw new Error(`Multiple tasks found under the heading: "${heading}".
The test design only supports one task per heading currently, so this is an error.

Edit "${task.path}" to move one of these lines to a separate heading:
"${acc[heading].originalMarkdown}"
"${task.originalMarkdown}"

And then rerun the command "Templater: Insert _meta/templates/convert_test_data_markdown_to_js.md".

For more info: https://publish.obsidian.md/tasks-contributing/Testing/Using+Obsidian+API+in+tests
`);
            }

            acc[heading] = task;
            return acc;
        }, {} as Record<string, Task>);
    });

    async function renderTask(task: Task, queryFilePath: string = 'query.md') {
        const renderer = makeHtmlResultsRenderer('', new TasksFile(queryFilePath));
        const queryRendererParameters = {
            allTasks: [task],
            allMarkdownFiles: [],
            backlinksClickHandler: () => Promise.resolve(),
            backlinksMousedownHandler: () => Promise.resolve(),
            editTaskPencilClickHandler: () => Promise.resolve(),
        };
        const container = document.createElement('div');

        await renderer.render(State.Warm, [task], container, queryRendererParameters);

        return container.querySelector('.task-description')?.innerHTML ?? '';
    }

    it('should not modify description when rendering in same file', async () => {
        const description = await renderTask(
            tasksByHeading['Basic Internal Links'],
            'Test Data/internal_heading_links.md',
        );
        expect(description).toMatchInlineSnapshot('"<span>#task Task with<br>[[#Basic Internal Links]]</span>"');
    });

    it('should convert internal heading links when rendering in different file', async () => {
        const description = await renderTask(tasksByHeading['Basic Internal Links'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with<br>[[Test Data/internal_heading_links.md#Basic Internal Links|Basic Internal Links]]</span>"',
        );
    });

    it('should handle multiple internal heading links in one description', async () => {
        const description = await renderTask(tasksByHeading['Multiple Links In One Task'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with<br>[[Test Data/internal_heading_links.md#Multiple Links In One Task|Multiple Links In One Task]] and<br>[[Test Data/internal_heading_links.md#Simple Headers|Simple Headers]]</span>"',
        );
    });

    it('should not modify regular file links', async () => {
        const description = await renderTask(tasksByHeading['External File Links'], 'query.md');
        expect(description).toMatchInlineSnapshot('"<span>#task Task with<br>[[Other File]]</span>"');
    });

    it('should handle header links with mixed link types', async () => {
        const description = await renderTask(tasksByHeading['Mixed Link Types'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with<br>[[Other File]] and<br>[[Test Data/internal_heading_links.md#Mixed Link Types|Mixed Link Types]]</span>"',
        );
    });

    it('should handle header links with file references', async () => {
        const description = await renderTask(tasksByHeading['Header Links With File Reference'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task<br>[[Test Data/internal_heading_links.md#Header Links With File Reference|Header Links With File Reference]] then<br>[[Other File#Some Header]] and<br>[[Test Data/internal_heading_links.md#Another Header|Another Header]]</span>"',
        );
    });

    it('should handle header links with special characters', async () => {
        const description = await renderTask(tasksByHeading['Headers-With_Special Characters'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with<br>[[Test Data/internal_heading_links.md#Headers-With_Special Characters|Headers-With_Special Characters]]</span>"',
        );
    });

    it('should handle links with aliases', async () => {
        const description = await renderTask(tasksByHeading['Aliased Links'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with<br>[[Test Data/internal_heading_links.md#Aliased Links|I am an alias]]</span>"',
        );
    });

    it('should not modify formatted text that looks like links in code blocks', async () => {
        const description = await renderTask(tasksByHeading['Links In Code Blocks'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with `[[#Links In Code Blocks]]` code block</span>"',
        );
    });

    it('should not modify escaped links', async () => {
        const description = await renderTask(tasksByHeading['Escaped Links'], 'query.md');
        expect(description).toMatchInlineSnapshot(
            '"<span>#task Task with \\[\\[#Escaped Links\\]\\] escaped link</span>"',
        );
    });
});
