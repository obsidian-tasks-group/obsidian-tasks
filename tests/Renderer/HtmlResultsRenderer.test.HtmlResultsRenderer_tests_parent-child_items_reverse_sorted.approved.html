<!--
    - [ ] parent 2
                - [ ] grand grand child
        - [ ] child 2
            - [ ] grandchild 1
        - [ ] child 1
    - [ ] parent 1
-->

<div>
  <button class="plugin-tasks-copy-button">Copy results</button>
  <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
    <li
      class="task-list-item plugin-tasks-list-item"
      data-task-priority="normal"
      data-task=""
      data-line="0"
      data-task-status-name="Todo"
      data-task-status-type="TODO">
      <input class="task-list-item-checkbox" type="checkbox" title="Right-click for options" data-line="0" />
      <span class="tasks-list-text">
        <span class="task-description"><span>parent 2</span></span>
      </span>
      <span class="task-extras">
        <span class="tasks-backlink">
          (
          <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
          )
        </span>
        <a class="tasks-edit" title="Edit task" href="#"></a>
      </span>
    </li>
    <li
      class="task-list-item plugin-tasks-list-item"
      data-task-priority="normal"
      data-task=""
      data-line="5"
      data-task-status-name="Todo"
      data-task-status-type="TODO">
      <input class="task-list-item-checkbox" type="checkbox" title="Right-click for options" data-line="5" />
      <span class="tasks-list-text">
        <span class="task-description"><span>parent 1</span></span>
      </span>
      <span class="task-extras">
        <span class="tasks-backlink">
          (
          <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
          )
        </span>
        <a class="tasks-edit" title="Edit task" href="#"></a>
      </span>
      <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
        <li
          class="task-list-item plugin-tasks-list-item"
          data-task-priority="normal"
          data-task=""
          data-line="0"
          data-task-status-name="Todo"
          data-task-status-type="TODO">
          <input class="task-list-item-checkbox" type="checkbox" title="Right-click for options" data-line="0" />
          <span class="tasks-list-text">
            <span class="task-description"><span>child 1</span></span>
          </span>
          <span class="task-extras">
            <span class="tasks-backlink">
              (
              <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
              )
            </span>
            <a class="tasks-edit" title="Edit task" href="#"></a>
          </span>
          <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
            <li
              class="task-list-item plugin-tasks-list-item"
              data-task-priority="normal"
              data-task=""
              data-line="0"
              data-task-status-name="Todo"
              data-task-status-type="TODO">
              <input class="task-list-item-checkbox" type="checkbox" title="Right-click for options" data-line="0" />
              <span class="tasks-list-text">
                <span class="task-description"><span>grandchild 1</span></span>
              </span>
              <span class="task-extras">
                <span class="tasks-backlink">
                  (
                  <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
                  )
                </span>
                <a class="tasks-edit" title="Edit task" href="#"></a>
              </span>
              <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
                <li><span>list item grand grand child</span></li>
              </ul>
            </li>
          </ul>
        </li>
        <li
          class="task-list-item plugin-tasks-list-item"
          data-task-priority="normal"
          data-task=""
          data-line="1"
          data-task-status-name="Todo"
          data-task-status-type="TODO">
          <input class="task-list-item-checkbox" type="checkbox" title="Right-click for options" data-line="1" />
          <span class="tasks-list-text">
            <span class="task-description"><span>child 2</span></span>
          </span>
          <span class="task-extras">
            <span class="tasks-backlink">
              (
              <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
              )
            </span>
            <a class="tasks-edit" title="Edit task" href="#"></a>
          </span>
          <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
            <li>
              <span>non task grandchild</span>
              <ul class="contains-task-list plugin-tasks-query-result tasks-layout-hide-urgency">
                <li
                  class="task-list-item plugin-tasks-list-item"
                  data-task-priority="normal"
                  data-task=""
                  data-line="0"
                  data-task-status-name="Todo"
                  data-task-status-type="TODO">
                  <input
                    class="task-list-item-checkbox"
                    type="checkbox"
                    title="Right-click for options"
                    data-line="0" />
                  <span class="tasks-list-text">
                    <span class="task-description"><span>grand grand child</span></span>
                  </span>
                  <span class="task-extras">
                    <span class="tasks-backlink">
                      (
                      <a rel="noopener" target="_blank" class="internal-link">inheritance_rendering_sample</a>
                      )
                    </span>
                    <a class="tasks-edit" title="Edit task" href="#"></a>
                  </span>
                </li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
  <div class="task-count">6 tasks</div>
</div>
